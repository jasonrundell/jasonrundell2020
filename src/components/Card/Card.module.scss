.root {
  display: inline-block;
  transition: all 0.3s cubic-bezier(0.25, 0.8, 0.25, 1);
  margin: $sizeNormal;
  border-radius: 0.125rem;
  box-shadow: 0 1px 3px rgba(0, 0, 0, 0.12), 0 1px 2px rgba(0, 0, 0, 0.24);
  background: #fff;
}

.elevation0 {
  box-shadow: none;
}

.elevation1 {
  box-shadow: 0 1px 3px rgba(0, 0, 0, 0.12), 0 1px 2px rgba(0, 0, 0, 0.24);
}

.elevation2 {
  box-shadow: 0 3px 6px rgba(0, 0, 0, 0.16), 0 3px 6px rgba(0, 0, 0, 0.23);
}

.elevation3 {
  box-shadow: 0 10px 20px rgba(0, 0, 0, 0.19), 0 6px 6px rgba(0, 0, 0, 0.23);
}

.elevation4 {
  box-shadow: 0 14px 28px rgba(0, 0, 0, 0.25), 0 10px 10px rgba(0, 0, 0, 0.22);
}

.elevation5 {
  box-shadow: 0 19px 38px rgba(0, 0, 0, 0.3), 0 15px 12px rgba(0, 0, 0, 0.22);
}

.animation {
  &:hover {
    box-shadow: 0 14px 28px rgba(0, 0, 0, 0.25), 0 10px 10px rgba(0, 0, 0, 0.22);
  }
}
