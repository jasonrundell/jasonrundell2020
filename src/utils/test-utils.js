export function catchWarnings() {
  const originalConsoleError = global.console.error
  beforeEach(() => {
    /**
     * I Want to know if values passed to props fail the type check
     * */
    global.console.error = (...args) => {
      const propTypeFailures = [/Warning:/] // e.g. [/Invalid prop/, /Failed prop type/]

      if (propTypeFailures.some(p => p.test(args[0]))) {
        throw new Error(args[0])
      }

      originalConsoleError(...args)
    }
  })
}

/**
 * Mix an match help from:
 * https://daveceddia.com/testing-for-proptypes-errors/
 * https://www.jackfranklin.co.uk/blog/failing-tests-on-react-proptypes/
 */
