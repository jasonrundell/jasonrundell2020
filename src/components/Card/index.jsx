import React from 'react'
import PropTypes from 'prop-types'

import { classNames } from '../../utils/css'

import { root } from './Card.module.scss'

const Card = ({ children, elevation, isElevationAnimated }) => {
  const classes = classNames(
    styles.root,
    styles[`elevation${elevation}`],
    isElevationAnimated && styles.animation
  )
  return <div className={classes}>{children}</div>
}

Card.defaultProps = {
  elevation: 2,
  isElevationAnimated: true,
}

Card.propTypes = {
  elevation: PropTypes.oneOf([0, 1, 2, 3, 4, 5]),
  isElevationAnimated: PropTypes.bool,
  children: PropTypes.any.isRequired,
}

export default Card
