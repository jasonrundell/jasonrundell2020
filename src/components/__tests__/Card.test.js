import React from 'react'
import { render, unmountComponentAtNode } from 'react-dom'
import { act } from 'react-dom/test-utils'
import { catchWarnings } from './test-utils'
// import { create } from 'react-test-renderer'

import Card from '../Card'

catchWarnings()

let container = null
beforeEach(() => {
  container = document.createElement('div')
  document.body.appendChild(container)
})

afterEach(() => {
  unmountComponentAtNode(container)
  container.remove()
  container = null
})

describe('Badge component', () => {
  it('Card has required children', () => {
    act(() => {
      render(<Card>Card component</Card>, container)
    })
    expect(container.textContent).toBe('Card component')
  })

  it('Card elevations work', () => {
    act(() => {
      render(<Card elevation={0}>Elevation 0</Card>, container)
    })
    expect(container.textContent).toBe('Elevation 0')

    act(() => {
      render(<Card elevation={1}>Elevation 1</Card>, container)
    })
    expect(container.textContent).toBe('Elevation 1')

    act(() => {
      render(<Card elevation={2}>Elevation 2</Card>, container)
    })
    expect(container.textContent).toBe('Elevation 2')

    act(() => {
      render(<Card elevation={3}>Elevation 3</Card>, container)
    })
    expect(container.textContent).toBe('Elevation 3')

    act(() => {
      render(<Card elevation={4}>Elevation 4</Card>, container)
    })
    expect(container.textContent).toBe('Elevation 4')

    act(() => {
      render(<Card elevation={5}>Elevation 5</Card>, container)
    })
    expect(container.textContent).toBe('Elevation 5')
  })
})
