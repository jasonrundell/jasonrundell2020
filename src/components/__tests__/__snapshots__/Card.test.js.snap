// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component Card renders correctly 1`] = `
<div
  className="root elevation2 animation"
>
  Card component
</div>
`;
